### ISSUE TITLE
SetReturnsDefault not working for Task<>

### ISSUE THREAD

**Issue by Albert221** [moq/moq4#673]
Explanation in comment.

```cs
public interface ITestInterface
{
    Task<bool> SomeMethod();
}

[Fact]
public async void TestBug()
{
    var mock = new Mock<ITestInterface>();
    mock
        .Setup(x => x.SomeMethod());

    mock.SetReturnsDefault(Task.FromResult(false));

    var obj = mock.Object;

    // `await` below throws NullReferenceException, because `obj.SomeMethod()` returns `null`
    Assert.False(await obj.SomeMethod());
}
```

Looks like `Mock.SetReturnsDefault()` doesn't like `Task` type, because `obj.SomeMethod()` returns `null` instead of `Task<bool>`. Although, if I add `.ReturnsAsync(false)` after `.Setup` (and remove `SetReturnsDefault` line) then everything is okay, obviously.

**Comment by stakx** [comment]
Hi @Albert221, there's no problem with `SetReturnsDefault` here that is specific to `Task<T>`. Instead, you are observing the somewhat non-intuitive way how Moq determines default return values for method calls.

* As soon as there is a setup for some method call (such as in your case), any return value configured by `SetReturnsDefault` are ignored. This mechanism works only for method calls that do not have any matching setups.

* If a method call cas a matching setup, but that setup doesn't explicitly specify the return value (either via `CallBase` or `Returns`), then the return value will be the default value for the method's return value&mdash;in C# parlance, `default(T)`. This is what you're observing, since `default(Task<bool>)` is `null`.

So, basically, if you just remove the dummy setup for `SomeMethod`, the remainder of your code will work as expected.

**Comment by Albert221** [comment]
Okay, it all makes sense now, thank you for your explanation!

I suggest adding this explanation (or reference to it) to XML documentation maybe? I've just searched for that in the Wiki, and it is present there, but I wonder whether it would be helpful and would save time for people like me :)

**Comment by stakx** [comment]
@Albert221: That's a good idea! I'd suggest that for now, we add a hint in the XML documentation for `Mock.SetReturnsDefault` saying that those default return values are used only in the absence of a setup for a method call.

Would you like to submit a PR?

**Comment by Albert221** [comment]
Will do!

**Comment by stakx** [comment]
@Albert221: Thank you for contributing to Moq! :+1:
